### PR TITLE
Using the CsWin32 PInvoke library

### PR DESCRIPTION
**Pull request by fernandanavamoya**
Converted these threads:
-GetProcAddress
-GetTickCount
-GetUserDefaultLCID
-GlobalSize

Added PInvoke.GetModuleHandle.cs file

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7499)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/NativeMethods.txt [hunk 1]
```diff
@@ -2,15 +2,27 @@ BitBlt
...
+GetProcAddress
 GetThreadLocale
+GetTickCount
 GlobalAlloc
 GlobalFree
 GlobalLock
 GlobalReAlloc
+GlobalSize
 GlobalUnlock
+GetProcAddress
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Primitives/src/NativeMethods.txt:23]: Sort order missing for new additions here.

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        public static unsafe HINSTANCE GetModuleHandle(
+            string lpModuleName)
+        {
+            // We are using GetModuleHandleEx because of the difficulties of the safe handle overload generated when
+            // using GetModuleHandle
+            fixed (char* lpModuleNameLocal = lpModuleName)
+            {
+                HINSTANCE hModule;
+                BOOL result = GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, lpModuleNameLocal, &hModule);
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs:19]: if you are not using/checking `result`,  you may want to ignore the return value here.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 3]
```diff
@@ -118,20 +119,28 @@ private static bool InitializeComCtlSupportsVisualStyles()
             //
             // GetModuleHandle  returns a handle to a mapped module without incrementing its
             // reference count.
-            IntPtr hModule = Kernel32.GetModuleHandleW(Libraries.Comctl32);
-            if (hModule != IntPtr.Zero)
+            var hModule = PInvoke.GetModuleHandle(Libraries.Comctl32);
+            string IpProcName = "ImageList_WriteEx";
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:123]: const string?

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 4]
```diff
@@ -118,20 +119,28 @@ private static bool InitializeComCtlSupportsVisualStyles()
...
+            nint ninthModule = Kernel32.LoadComctl32(StartupPath);
+            if (ninthModule == 0)
             {
                 return false;
             }
 
-            return Kernel32.GetProcAddress(hModule, "ImageList_WriteEx") != IntPtr.Zero;
+            fixed (byte* ptr = &test[0])
+            {
+                return PInvoke.GetProcAddress(hModule, (Foundation.PCSTR)ptr) != 0;
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:142]: isn't this `ninthModule`

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs [hunk 5]
```diff
@@ -3232,7 +3232,7 @@ public unsafe bool HasPropertyPages()
 
         private unsafe void ShowPropertyPageForDispid(Ole32.DispatchID dispid, Guid guid)
         {
-            nint pUnk = Marshal.GetIUnknownForObject(GetOcx());
+            IntPtr pUnk = Marshal.GetIUnknownForObject(GetOcx());
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:3235]: why this change?

File: src/System.Windows.Forms/src/System/Windows/Forms/NativeWindow.cs [hunk 6]
```diff
@@ -163,20 +164,24 @@ internal unsafe void ForceExitMessageLoop()
...
         /// </summary>
-        internal static IntPtr DefaultWindowProc
+        internal unsafe static nint DefaultWindowProc
         {
             get
             {
-                if (s_defaultWindowProc == IntPtr.Zero)
+                var hModule = PInvoke.GetModuleHandle(Libraries.Comctl32);
+                string IpProcName = "DefWindowProcW";
+                byte[] test = Encoding.ASCII.GetBytes(IpProcName);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/NativeWindow.cs:173]: if `s_defaultWindowProc `is not zero, we should avoid these calls and bail out as it was doing earlier.

File: src/System.Windows.Forms.Primitives/src/NativeMethods.txt (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/NativeMethods.txt:23]: This is a dup - likely a bad merge

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Windows.Win32.Foundation;
+
+namespace Windows.Win32
+{
+    internal static partial class PInvoke
+    {
+        public static unsafe HINSTANCE GetModuleHandle(
+            string lpModuleName)
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs:12]: Keep on the same line

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 3 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:123]: Incorrect casing too.
Ditto below

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 8]
```diff
@@ -118,20 +119,28 @@ private static bool InitializeComCtlSupportsVisualStyles()
...
-            if (hModule != IntPtr.Zero)
+            var hModule = PInvoke.GetModuleHandle(Libraries.Comctl32);
+            string IpProcName = "ImageList_WriteEx";
+            byte[] test = Encoding.ASCII.GetBytes(IpProcName);
+            fixed (byte* ptr = &test[0])
             {
-                return Kernel32.GetProcAddress(hModule, "ImageList_WriteEx") != IntPtr.Zero;
+                if (hModule != 0)
+                {
+                    return PInvoke.GetProcAddress(hModule,(Foundation.PCSTR)ptr) != 0;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:129]: Incorrect spacing

File: src/System.Windows.Forms.Primitives/src/NativeMethods.txt (hunk 1 above)
Comment (fernandanavamoya) [reply src/System.Windows.Forms.Primitives/src/NativeMethods.txt]: You're right, already corrected it.

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs (hunk 5 above)
Comment (fernandanavamoya) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:3235]: In the last coding review, it was suggested that `IntPtr` is better for representing an opaque value like a pointer.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 4 above)
Comment (fernandanavamoya) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:142]: `ninthModule` is a `nint` which stores the return value of LoadComctl32, whereas `hModule` is an `HINSTANCE` which stores the return value of GetModuleHandle. The method `GetProcAddres` has as attributes `(HINSTANCE hModule, PCSTR IpProcName)`, this is why in line 142 it's `hModule` instead of `ninthModule`.

[https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getprocaddress](url)

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 8 above)
Comment (fernandanavamoya) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:129]: By incorrect spacing do you mean adding a space between the `,` and `(Foundation.PCSTR)ptr`?

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 4 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:142]: Thank you for explanation here.

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:3235]: I'd leave it as `nint`. When we really want to reflect pointers we should be using `void*`, etc. and unsafe. That isn't necessary here as we're not currently changing it.

Oleaut32.OCPFIPARAMS should be using `void*` for `ppUnk`. Not necessary to fix unless we're actually making changes to it.

File: src/System.Windows.Forms/src/System/Windows/Forms/NativeWindow.cs [hunk 9]
```diff
@@ -7,8 +7,9 @@
 using System.Globalization;
 using System.Runtime.ConstrainedExecution;
 using System.Runtime.InteropServices;
+using System.Text;
```
Comment (elachlan) [inline src/System.Windows.Forms/src/System/Windows/Forms/NativeWindow.cs:10]: This using is breaking the test.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 8 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Yes.
image

File: src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Windows/Win32/PInvoke.GetModuleHandle.cs:18]: We must validate the return, and thow 
image

File: src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs (hunk 5 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/AxHost.cs:3235]: This is still outstanding

**Comment by RussKie** [comment]
Why was this merged? There are outstanding unresolved feedback issues.
